integration/tests/cook/util.py

```diff
@@ -13,20 +14,108 @@
 session.headers['User-Agent'] = f"Cook-Scheduler-Integration-Tests ({session.headers['User-Agent']})"
 
-DEFAULT_TEST_TIMEOUT_SECS = 600 # no individual test exceeds 10 minutes
+# default time limit for each individual integration test
+# if a test takes more than 10 minutes, it's probably broken
+DEFAULT_TEST_TIMEOUT_SECS = 600
 
+# default time limit used by most wait_* utility functions
+# 2 minutes should be more than sufficient on most cases
 DEFAULT_TIMEOUT_MS = 120000
 
+# Cached data from Cook's /info endpoint
+_cook_info = None
+
 
 def continuous_integration():
-    # Travis-CI sets this env variable automatically
+    """Returns true if the CONTINUOUS_INTEGRATION environment variable is set, as done by Travis-CI."""
     return os.environ.get('CONTINUOUS_INTEGRATION')
 
 
 def has_docker_service():
+    """Returns true if docker services appear to be available to the testing environment."""
     return os.path.exists('/var/run/docker.sock')
 
 
+class User(object):
+    """
+    Object representing a Cook user, which holds authentication details.
+    User objects can be used with python's `with` blocks
+    to conveniently set a user for a sequence of commands.
+    """
+
+    def __init__(self, name):
+        self.name = name
+        self.auth = (name, '')
+        self.previous_auth = None
+
+    def __enter__(self):
+        global session
+        logger.debug(f"Switching to user {self.name}")
+        assert self.previous_auth is None
+        self.previous_auth = session.auth
+        session.auth = self.auth
+
+    def __exit__(self, ex_type, ex_val, ex_trace):
+        global session
+        logger.debug(f"Switching back from user {self.name}")
+        assert self.previous_auth is not None
+        session.auth = self.previous_auth
+        self.previous_auth = None
+
+
+class UserFactory(object):
+    """
+    Factory object used to create unique user names for a given test function.
+    Usernames are composed of the test name and and increasing integer values.
+    """
+
+    def __init__(self, test_handle):
+        test_id = test_handle.id()
+        base_name = test_id[test_id.rindex('.test_')+6:].lower()
+        self.__user_generator = ( f'{base_name}_{i}' for i in range(1000000) )
+
+    def new_user(self):
+        """Return a fresh User object."""
+        return User(next(self.__user_generator))
+
+    def new_users(self, count=None):
+        """Return a sequence of `count` fresh User objects."""
+        return map(User, itertools.islice(self.__user_generator, 0, count))
+
+
+# Default user for multi-user test runs
+default_user = User('root')
+
+
+def multi_cluster_tests_enabled():
+    """
+    Returns true if the CONTINUOUS_INTEGRATION environment variable is set,
+    indicating that multiple cook scheduler instances are running.
+    """
+    return os.getenv('COOK_MULTI_CLUSTER') is not None
+
+
+def _cook_auth_scheme():
```

Comment:
Instead of caching `_cook_info` in a global, can you just memoize this function call?